lib/vrl/stdlib/src/match_datadog_query.rs

```diff
@@ -0,0 +1,1870 @@
+                result: Ok("true"),
+            },
+        ]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+        let query_value = arguments.required_literal("query")?.to_value();
+
+        // Query should always be a string.
+        let query = query_value
+            .try_bytes_utf8_lossy()
+            .expect("datadog search query not bytes");
+
+        // Compile the Datadog search query to AST.
+        let node = parse(&query).map_err(|e| {
+            Box::new(ExpressionError::from(e.to_string())) as Box<dyn DiagnosticError>
+        })?;
+
+        Ok(Box::new(MatchDatadogQueryFn { value, node }))
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::OBJECT,
+                required: true,
+            },
+            Parameter {
+                keyword: "query",
+                kind: kind::BYTES,
+                required: true,
+            },
+        ]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct MatchDatadogQueryFn {
+    value: Box<dyn Expression>,
+    node: QueryNode,
+}
+
+impl Expression for MatchDatadogQueryFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?.try_object()?;
+
+        Ok(matches_vrl_object(&self.node, &Value::Object(value)).into())
+    }
+
+    fn type_def(&self, _state: &state::Compiler) -> TypeDef {
+        type_def()
+    }
+}
+
+fn type_def() -> TypeDef {
+    TypeDef::new().infallible().boolean()
+}
+
+/// Match the parsed node against the provided VRL `Value`, per the query type.
+fn matches_vrl_object(node: &QueryNode, obj: &Value) -> bool {
+    match node {
+        QueryNode::MatchNoDocs => false,
+        QueryNode::MatchAllDocs => true,
+        QueryNode::AttributeExists { attr } => exists(attr, obj),
+        QueryNode::AttributeMissing { attr } => !exists(attr, obj),
+        QueryNode::AttributeTerm { attr, value }
+        | QueryNode::QuotedAttribute {
+            attr,
+            phrase: value,
+        } => equals(attr, obj, value),
+        QueryNode::AttributeComparison {
+            attr,
+            comparator,
+            value,
+        } => match_compare(attr, obj, comparator, value),
+        QueryNode::AttributePrefix { attr, prefix } => {
+            match_wildcard_with_prefix(attr, obj, prefix)
+        }
+        QueryNode::AttributeWildcard { attr, wildcard } => match_wildcard(attr, obj, wildcard),
+        QueryNode::AttributeRange {
+            attr,
+            lower,
+            lower_inclusive,
+            upper,
+            upper_inclusive,
+        } => range(attr, obj, lower, *lower_inclusive, upper, *upper_inclusive),
+        QueryNode::NegatedNode { node } => !matches_vrl_object(node, obj),
+        QueryNode::Boolean { oper, nodes } => {
+            if let BooleanType::And = oper {
```

Comment:
[nit] maybe 'match' would be less error-prone in case we add some other boolean operators in the future?

Reply:
IIRC, clippy balks when using match as a two-branch if/else.

Reply:
ah, interesting, I'd personally prefer match in this case, but that's fine then 👌

Reply:
> IIRC, clippy balks when using match as a two-branch if/else.

I am not convinced that's true. It _does_ complain when you try to match against a boolean, but e.g. `Result` and `Option` both have two branches, and it doesn't complain there.

Still though, I'm fine with either approach.

Reply:
You're right. I remember seeing it for bool values. `match` is nicer.